### ISSUE TITLE
severe parsing bug in AS 3.10.0

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#306]
In GitLab by @etiennesky on Feb 16, 2018, 15:55

When running the LOCAL_SETUP job (sources/runtime/autosubmit/compilation.sh) on MN4 with recent AS version (3.10.0).

I encountered a bug in the AS parsing, running expid a0ts with ORCA025 nemo config

This statement

NEMO_CONFIG=%NEMO_resolution%_%ICE%

was translated to

NEMO_CONFIG=%NEMO_resolutionICE%

while it should have been translated to

NEMO_CONFIG=ORCA025L75_LIM3

You can see the log at a0ts_REMOTE_SETUP.20180216153747.err whereas after reverting back to AS 3.9.1 this log shows parsing was ok a0ts_REMOTE_SETUP.20180216155010.err .

I suggest you

1. revert back to default 3.9.1 as the default AS version in the modules
2. fix the bug and run 1 full testing suite of EC-Earth, as there well may be other bugs
3. make sure someone from CES runs a full test of auto-ecearth3 before any AS release

@ lbatista @dmanubens @pablohe @mcastril

**Comment by kinow** [comment]
In GitLab by @ lbatista on Feb 16, 2018, 16:02

Hi @etiennesky, actually we only made the version default **after** @pablohe encountered the very same problem #302 and we had discussed a workaround/temporary solution for this problem.

But I definitely agree with the 3rd point of running a full test of auto-ecearth3 before any AS release. 
In the mean time, if everybody agrees, I can revert the default version

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 16, 2018, 16:05

Hi I think it is a good practice to either let people know of the workaround (in the issue as well as in other manner), especially if the bug is present in the default version. 

Can you please write the workaround here? Thanks!

**Comment by kinow** [comment]
In GitLab by @ lbatista on Feb 16, 2018, 16:09

Hi @etiennesky sorry if it was not clear, but the workaround was added to the current version 3.10.0 (we rebuilt it before making it default), meaning that there is nothing extra that should be done by users (otherwise I would have sent an email letting everyone know). @pablohe gave me the ok that the problem had been fixed, hence why we made the version default. 

We acknowledge that the current parsing strategy is not the best, and we are working on coming up with a new (global) solution, probably by using a new library.

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 16, 2018, 16:10

Hi, sorry I misunderstood! 

I had the problem today after switching to AS 3.10.0 on besesautosubmit02, so the problem seems to still exist? Maybe I need to re-load the module then! I will try again!

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 16, 2018, 16:19

I re-loaded the AS module and same problem. maybe a deployment issue? 

I even tried to unload the module, refresh the module cache (with module refresh), same problem.

**Comment by kinow** [comment]
In GitLab by @pablohe on Feb 16, 2018, 16:22

I made a quick fix in branches/development/2017/r4737-bsc_trunk / runtime/autosubmit/compilation.sh

```
...
NEMO_resolution=%NEMO_resolution%
ICE=%ICE%
NEMO_CONFIG=${NEMO_resolution}_${ICE}
...
```

**Comment by kinow** [comment]
In GitLab by @ lbatista on Feb 16, 2018, 16:25

Hi @etiennesky, sorry it wasn't clear for me that when @pablohe gave me the ok, I didn't know he had slightly modified the script as well. In this case, maybe it's best that we report this to everybody along with some sort of "best practice" to avoid such errors in the mean time (what do you think? @dmanubens @pablohe @mcastril)

We will make this a priority as since the modification introduced due to #289, we have been worried about the possible drawbacks of simply "appending" this to the existing solution. Clearly it has already backfired and we believe a more robust change is necessary.

Anyway, I appreciate the suggestions of running a full test before any deployment/release, I agree it's really important.

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 16, 2018, 16:25

@pablohe easier workaround was to revert to AS 3.9.1 no?

If it was not fixed in AS then please revert the default version back to 3.9.1 . 

There might be other templates which might be affected by the bug.

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 16, 2018, 16:26

Hi @ lbatista no problem these things happen. thanks for your help.

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 16, 2018, 16:34

@pablohe I think you need to be a bit more defensive and not assume that everyone updates their branches or experiments from bsc_trunk on a daily basis. 

If there is a bug in the default version of autosubmit users should be aware of that and it should be fixed or reverted back to the previous version. Thanks!

**Comment by kinow** [comment]
In GitLab by @pablohe on Feb 16, 2018, 16:35

If you use bsc_trunk/bsc_nemo is not posible. I have adapted the tests to use the new autosubmit (see https://earth.bsc.es/gitlab/es/auto-ecearth3/commit/2039b2b3b6a1942f6c9932b6b94a6c1f4fdd57e6)

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 16, 2018, 16:36

Yes, the logical fix is to go back to 3.9.1 . Usually new versions are on testing for some weeks until it's time to consider it stable.

So @pablohe will have to turn again back the comparisons against the "empty var". %VAR% == TRUE || %VAR%. 

We did the change to 3.10 because of that, but having this bug we have to recover the usual workflow.

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 16, 2018, 16:41

I understand now. I think that all experiments will fail in random places when using AS 3.10.0.

So my recommendation would be

1. for anything older than current bsc_trunk, use AS 3.9.1 or older
2. for bsc_trunk use AS 3.10.0 - but let people know!

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 16, 2018, 16:45

That would take less, but at the same time I think is more robust to put again back all that is not compatible with AS 3.9.0, and would wait to apply the commits back when 3.10.0 is stable.

We still don't know if there is any bug around. And we will use these new branches (bsc_nemo) to run production experiments, as EnKF.

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 16, 2018, 17:16

Thanks everyone!

The strategy adopted is to revert default to 3.9.1 (already done), and to make sure that bsc_trunk and bsc_nemo are compatible with 3.9.1 and 3.10.0, but for now officially recommending 3.10.0.

Agreed?

**Comment by kinow** [comment]
In GitLab by @etiennesky on Mar 13, 2018, 13:34

@ rfernand had this problem as well

**Comment by kinow** [comment]
In GitLab by @etiennesky on Mar 13, 2018, 13:40

more info... @ rfernand was using AS 3.10.0 with a branch that had not yet been updated. This will probably happen when people use 3.10.0 with auto-ecearth 3.2.2 .

Any idea on when this bug will be resolved? thanks!

**Comment by kinow** [comment]
In GitLab by @ rfernand on Mar 13, 2018, 16:18

Hi,

My exp is a0xq and I am using the branch 'bsc_trunk_saveini' and running auto-ecearth 3.2.3.

**Comment by kinow** [comment]
In GitLab by @ lbatista on Mar 13, 2018, 17:04

Hi,

Any script which has `if [[ "%%_%'Reference to deleted milestone  == '%_%" ]]` will have this problem (if there are problems with any other script, please report), but regarding the auto-ecearth branches, I do not know which ones have been updated.

As a general note, **you should expect**:
1. **Anything** between %% in the template to be considered a variable
2. Non initialized variables to be replaced by a blank value ''

This is the expected behavior from 3.10, after it was requested to replace non-initialized variables by blank (point 1 has been around for a while). 

As an improvement to our parsing strategy, in order to avoid this type of problems of wrongly defined variables, we have already implemented a fix: replacing by blank only after all the actual/defined variables have been replaced in the template. This is already done, but the update has not been propagated to 3.10 yet. Tbd whether to change 3.10 or to create 3.10.1 also with few other features/fixes.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 20, 2018, 16:12

Instead of updating old Auto-EC-Earth branches to new Autosubmit versions, specially when they bring a compatibility break, we rather could check what Autosubmit version is being used in the LOCAL_SETUP and stop if the version used is bigger than the one supported by the Auto-EC-Earth branch... "autosubmit -v" should work from inside the LOCAL_SETUP right?

If we update old branches to newer Autosubmit we are in risk of breaking compatibility with the experiments that where created using an specific Autosubmit version (that info is in the experiment's conf files).

@ lbatista , about the fix, what would happen if `%undefined%_%undefined%` appears in the code? I guess %_% could again be replaced first? I still think that the way to avoid this issues is to replace occurrences from left to right. Maybe you are already taking this into consideration.

**Comment by kinow** [comment]
In GitLab by @ lbatista on Mar 20, 2018, 16:31

@mcastril remember that `%_%` is only detected as a variable if `if [[ "%%_%'Reference to deleted milestone  == '%_%" ]]` is in the template. Otherwise it is not a problem. 

In any case, yes, what happens when there is `%undefined1%_%undefined2%` in the template is that undefined1 and undefined2 will be placed into a list (therefore we can assure the order). Then the values on this list are only replaced in the template after all the defined variables are already there

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 20, 2018, 16:55

Thanks @ lbatista ,

I was pointing out to the case that `%_%` is in the template, as also as `%undefined%_%undefined%`. I was wondering if replacing defined variables before, would ensure that my expresion is not translated to `%undefinedundefined%`. That's why I was commenting about ensuring the order.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 5, 2019, 11:53

closed
